### PR TITLE
Remove symbol mutation utils

### PR DESCRIPTION
**Pull request by mfussenegger**
See https://github.com/crate/crate/pull/6175 for the reasons

### CODE REVIEW

File: sql/src/main/java/io/crate/analyze/symbol/Function.java [hunk 1]
```diff
@@ -39,10 +39,6 @@ public Function(FunctionInfo info, List<Symbol> arguments) {
         return arguments;
     }
 
-    public void setArgument(int index, Symbol symbol) {
-        arguments.set(index, symbol);
-    }
-
```
Comment (mxm) [inline sql/src/main/java/io/crate/analyze/symbol/Function.java]: 👍

File: sql/src/main/java/io/crate/analyze/symbol/FunctionCopyVisitor.java [hunk 2]
```diff
@@ -41,6 +44,9 @@ protected Function processAndMaybeCopy(Function func, C context) {
         List<Symbol> args = func.arguments();
         switch (args.size()) {
             // specialized functions to avoid allocations for common cases
+            case 0:
+                return func;
```
Comment (mxm) [inline sql/src/main/java/io/crate/analyze/symbol/FunctionCopyVisitor.java]: 😏

File: sql/src/main/java/io/crate/analyze/symbol/FunctionCopyVisitor.java [hunk 3]
```diff
@@ -94,7 +100,25 @@ private Function oneArg(Function func, C context) {
...
+    public Symbol visitDynamicReference(DynamicReference ref, C context) {
+        return visitReference(ref, context);
+    }
+
+    @Override
     protected Symbol visitSymbol(Symbol symbol, C context) {
         return symbol;
     }
+
+    public List<Symbol> process(Collection<? extends Symbol> symbols, C context) {
```
Comment (mxm) [inline sql/src/main/java/io/crate/analyze/symbol/FunctionCopyVisitor.java]: `processCollection(Collection<? extends Symbol> symbols, C context) {`?

**Comment by mfussenegger** [comment]
Regarding benchmarks: Unfortunately we currently hardly have any JMH benchmarks which contain any larger function expressions. And in the full-stack benchmarks we're not able to detect any changes in the nanosecond or microsecond range as the variance is too high.

**Pull request merged**